protocol/contracts/libraries/Silo/LibSilo.sol

```diff
@@ -8,98 +8,241 @@ pragma experimental ABIEncoderV2;
+        if (s.s.roots == 0) {
+            roots = uint256(stalk.mul(C.getRootsBase()));
+        } else  {
+            roots = s.s.roots.mul(stalk).div(s.s.stalk);
+            if (block.number - s.season.sunriseBlock <= 25) {
+                uint256 rootsWithoutEarned = s.s.roots.add(s.newEarnedRoots).mul(stalk).div(s.s.stalk - (s.newEarnedStalk));
+                uint256 deltaRoots = rootsWithoutEarned - roots;
+                s.newEarnedRoots = s.newEarnedRoots.add(uint128(deltaRoots));
+                s.a[account].deltaRoots = uint128(deltaRoots);
+            } 
+        }
 
+        // increment user and total stalk
         s.s.stalk = s.s.stalk.add(stalk);
         s.a[account].s.stalk = s.a[account].s.stalk.add(stalk);
 
+        // increment user and total roots
         s.s.roots = s.s.roots.add(roots);
         s.a[account].roots = s.a[account].roots.add(roots);
+
         emit StalkBalanceChanged(account, int256(stalk), int256(roots));
     }
 
-    function decrementBalanceOfSeeds(address account, uint256 seeds) private {
+    //////////////////////// BURN ////////////////////////
+
+    /**
+     * @dev WRAPPER: Burns Seeds, Stalk and Roots from `account`.
+     */
+    function burnSeedsAndStalk(
+        address account,
+        uint256 seeds,
+        uint256 stalk
+    ) internal {
+        burnSeeds(account, seeds);
+        burnStalk(account, stalk); // also burns Roots
+    }
+    
+    /**
+     * @dev Burns Seeds from `account`.
+     */
+    function burnSeeds(address account, uint256 seeds) private {
         AppStorage storage s = LibAppStorage.diamondStorage();
+
+        // Decrease supply of Seeds; Remove Seeds from the balance of `account`
         s.s.seeds = s.s.seeds.sub(seeds);
         s.a[account].s.seeds = s.a[account].s.seeds.sub(seeds);
+
         emit SeedsBalanceChanged(account, -int256(seeds));
     }
 
-    function decrementBalanceOfStalk(address account, uint256 stalk) private {
+    /**
+     * @dev Burns Stalk and Roots from `account`.
+     *
+     * For an explanation of Roots accounting, see {FIXME(doc)}.
+     */
+    function burnStalk(address account, uint256 stalk) private {
         AppStorage storage s = LibAppStorage.diamondStorage();
         if (stalk == 0) return;
 
-        uint256 roots = s.s.roots.mul(stalk).div(s.s.stalk);
+       
+        uint256 roots;
+        // Calculate the amount of Roots for the given amount of Stalk.
+        // We round up as it prevents an account having roots but no stalk.
+        
+        // if the user withdraws in the same block as sunrise, they forfeit their earned beans for that season
+        // this is distrubuted to the other users.
+        // should this be the same as the vesting period?
+        if(block.number - s.season.sunriseBlock <= 25){
+            roots = s.s.roots.mulDiv(
+            stalk,
+            s.s.stalk-s.newEarnedStalk,
+            LibPRBMath.Rounding.Up);
```

Comment:
@Brean0 previous logic was using safe math that rounds towards zero. why the change here and what are the implications?

Reply:
previously the roots here was truncated - with the introduction of more complex math, this adds an additional fail safe to prevent rounding errors. This follows defi standards (we round down when issuing roots, round up when removing roots) to prevent the protocol from being undercollateralized

Reply:
Additionally, line 95 adds an additional precaution to prevent underflow.

Reply:
what is the precaution? line 95 is `mintStalk(account, stalk); // also mints Roots`

Reply:
let's talk about rounding on the next call. i think i am missing a lot of context